File: 494_target_sum.py
class Solution(object):
    def findTargetSumWays(self, nums, target):
        memo = {}
        def dfs(i, t):
            if i == len(nums):
                return 1 if t == 0 else 0
            if (i, t) in memo:
                return memo[(i, t)]
            memo[(i, t)] = dfs(i + 1, t - nums[i]) + dfs(i + 1, t + nums[i])
            return memo[(i, t)]
        return dfs(0, target)

File: 494_target_sum.cpp
class Solution {
public:
    int ewb(int m, vector<int>& n, int o, int p) {
        return m == p ? o == 0 : ewb(m + 1, n, o - n[m], p) + ewb(m + 1, n, o + n[m], p);
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        return ewb(0, nums, target, nums.size());
    }
};
